### DISCUSSION TITLE
MicroPython on ESP32-CAM using SD

### DISCUSSION THREAD

**Discussion by moferreira** [micropython/micropython#15965] [ESP32] [answered]
Using sdcard.py (via "import sdcard") is it possible to configure the SD card adapter to work as  SPI? If so, could you help me?

ESP32-CAM pin out it depicted below. No camera is connected to it.
![Screenshot from 2024-10-04 23-41-04](https://github.com/user-attachments/assets/72339a1b-9ba6-4bd6-bb7e-101ca926aafa)

**Comment by moferreira** [comment]
Sorry... Should anyone get the same problem, here is the solution.
```
import machine
import os
import sdcard
spi = machine.SPI(1, baudrate=1000000, polarity=0, phase=0,
                  sck=machine.Pin(14), mosi=machine.Pin(15), miso=machine.Pin(2))
cs = machine.Pin(13, machine.Pin.OUT)
sd = sdcard.SDCard(spi, cs)
try:
    vfs = os.VfsFat(sd)
    os.mount(vfs, '/sd')
    print("SD card mounted successfully")
except Exception as e:
    print("Failed to mount SD card:", e)
print("SD card contents:", os.listdir('/sd'))
```

**Comment by shariltumin** [comment]
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/sdmmc_host.html

You can use ESP-SDMMC instead of the SPI interface, here's how.

Slot 1
CMD GPIO15
CLK GPIO14
D0  GPIO02
D1  GPIO04
D1  GPIO12
D3  GPIO13

At high speed frq = 40000000  you might (will) get 
E (1500539) sdmmc_cmd: sdmmc_read_sectors_dma: sdmmc_send_cmd returned 0x109
so keep it at 20000000

A simple test script:

```py
import machine, os
from time import sleep

slt = 1 # slot 0 use by flash memory (1: default)
wid = 1 # sdmmc (sdio) - 1 bit at time (default). Either 1 or 4
        # wid=1: Use GPIO: 15,14,2 (save 3 Pins)
        # wid=4: Use GPIO: 15,14,2,4,12,13
        #        Data1@GPIO4, you will get flashing LED-flash
frq = 20000000 # defult:20000000 (or 10000000), high speed 40000000 (not stable)

sd=machine.SDCard(slot=slt, width=wid, freq=frq)

os.mount(sd, "/sd")

for i in range(100): # 100 frames video
    print(gc.mem_free())
    data = b'sfsdfsdfeqeqw12414asfasf33qrfsafq3r332rfsfsdff3r3232rgsdvdvwr3efes'*500
    if len(data) > 0:
       print('Frame:', i)
       fn = '/sd/data-%03d.bin'%i
       f = open(fn, 'wb')
       w=f.write(data)
       print(i,w,len(data))
       f.close()
       del f
    sleep(0.5)

os.listdir("/sd")
os.umount("/sd")
```

**Comment by moferreira** [comment]
My problem is actually solved. However, your solution may be very, very helpful if we need speed. This said, on a ESP32-CAM board I get:

```
>>> import machine, os
>>> from time import sleep
>>> 
>>> slt = 1 # slot 0 use by flash memory (1: default)
>>> wid = 1 # sdmmc (sdio) - 1 bit at time (default). Either 1 or 4
>>>         # wid=1: Use GPIO: 15,14,2 (save 3 Pins)
>>>         # wid=4: Use GPIO: 15,14,2,4,12,13
>>>         #        Data1@GPIO4, you will get flashing LED-flash
>>> frq = 10000000 # defult:20000000 (or 10000000), high speed 40000000 (not stable)
>>> 
>>> sd=machine.SDCard(slot=slt, width=wid, freq=frq)
>>> 
>>> os.mount(sd, "/sd")
E (4328) sdmmc_common: sdmmc_init_ocr: send_op_cond (1) returned 0x107
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: 16

```

**Reply by moferreira** [reply]
![Screenshot from 2024-10-04 23-02-39](https://github.com/user-attachments/assets/eee525f3-bf48-4350-aba6-c98daf9c14b6)

**Reply by shariltumin** [reply, marked as answer]
Try to unplug the board from the power source and plug it back in again. machine.reset() does not seem to work in this case.

**Reply by moferreira** [reply]
Thank you! You look extremely knowledgeable on MicroPython... I was just to open a new question about if MicroPython has a Wireguard implementation. This said, should I open it or you know it does not implement, yet, Wireguard? If it implements, I preffer to post the question so everyone get to know how to do it.

**Reply by shariltumin** [reply]
This I do not know. Please post a new question, maybe other can help you with that. I'm glad that you will consider MicroPython for your projects, you will save a lot of time and frustration MCU are becoming more powerful and cheaper, we have no need to use C or C++ anymore for IoT works. Best of luck
